blocks/graph/__init__.py

```diff
@@ -104,7 +104,15 @@ def auxiliary_variables(self):
     @property
     def scan_variables(self):
         """Variables of Scan ops."""
-        return list(chain(*[g.variables for g in self._scan_graphs]))
+        # BFS
+        scan_graphs = self._scan_graphs
+        var_list = []
+        while scan_graphs:
+            g = scan_graphs.pop(0)
+            var_list.append(g.variables)
+            if g._scan_graphs:
+                scan_graphs.extend(g._scan_graphs)
+        return list(chain(*var_list))
```

Comment:
This is a desirable change, but I have a few concerns:
 - modifying the list that you iterate over is hard to understand and can very likely cause errors. Please try to  simplify the code
- a test is required